src/python/WMComponent/RucioInjector/Database/MySQL/GetCompletedBlocks.py

```diff
@@ -0,0 +1,114 @@
+"""
+_GetCompletedBlocks_
+
+MySQL implementation of RucioInjector.GetClompletedBlocks
+
+Retrieve a list of blocks that are complete,
+including their location and the sites they
+are subscribed too
+
+"""
+
+from __future__ import division
+from __future__ import print_function
+
+from WMCore.Database.DBFormatter import DBFormatter
+
+
+class GetCompletedBlocks(DBFormatter):
+    """
+    Retrieves a list of blocks that are closed but sure if deletedable:
+      - workflow completions for all files in the block to be checked on a later stage
+      - subscription made for dataset is copy+delete
+      - subscription has been made at the Data Management system
+      - blocks hasn't been deleted yet
+    """
+    sql = """SELECT dbsbuffer_block.blockname,
+                    dbsbuffer_location.pnn,
+                    dbsbuffer_dataset.path,
+                    dbsbuffer_dataset_subscription.site,
+                    dbsbuffer_workflow.name,
+                    dbsbuffer_block.create_time
+             FROM dbsbuffer_dataset_subscription
+             INNER JOIN dbsbuffer_dataset ON
+               dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
+             INNER JOIN dbsbuffer_block ON
+               dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
+             INNER JOIN dbsbuffer_file ON
+               dbsbuffer_file.block_id = dbsbuffer_block.id
+             INNER JOIN dbsbuffer_workflow ON
+               dbsbuffer_workflow.id = dbsbuffer_file.workflow
+             INNER JOIN dbsbuffer_location ON
+               dbsbuffer_location.id = dbsbuffer_block.location
+             WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
+             AND dbsbuffer_dataset_subscription.subscribed = 1
+             AND dbsbuffer_block.status = 'Closed'
+             AND dbsbuffer_block.deleted = 0
+             GROUP BY dbsbuffer_block.blockname,
+                      dbsbuffer_location.pnn,
+                      dbsbuffer_dataset.path,
+                      dbsbuffer_dataset_subscription.site,
+                      dbsbuffer_workflow.name,
+                      dbsbuffer_block.create_time
+             HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed)
```

Comment:
> the amount of dbsbuffer blocks matching all these constraints must be equal to the amount of completed workflows in dbsbuffer. Is that correct?

Yes

> And a completed workflow in dbsbuffer comes from the fact that all work units (wmbs records) have been processed in that workflow, right?

Again, Yes.

My understanding here is: This basically assures all the records returned would include only blocks related  to workflows marked as completed in `dbsbuffer_workflow` table.  This one I did not change it is from the old DAO. So just to be 100%  sure we are not messing things here I'd like to hear @hufnagel 's opinion as well.

Reply:
Yes, I think you are right! But given that we already join dbsbuffer_workflow, I wonder why not having and extra AND clause in the WHERE block with this constraint `dbsbuffer_workflow.completed=1` (or whatever value it's supposed to have).

Reply:
Thanks for reiterating through this once  again @amaltaro 

Lets talk with examples here
Here is what this query returns as it is right now for a completely finished replay. 
```
SQL> SELECT
  2         count(*),
  3         SUM(dbsbuffer_workflow.completed),
  4         dbsbuffer_block.blockname,
  5         dbsbuffer_location.pnn,
  6         dbsbuffer_dataset.path,
  7         dbsbuffer_dataset_subscription.site,
  8         dbsbuffer_workflow.name
  9  FROM dbsbuffer_dataset_subscription
 10  INNER JOIN dbsbuffer_dataset ON
 11    dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 12  INNER JOIN dbsbuffer_block ON
 13    dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 14  INNER JOIN dbsbuffer_file ON
 15    dbsbuffer_file.block_id = dbsbuffer_block.id
 16  INNER JOIN dbsbuffer_workflow ON
 17    dbsbuffer_workflow.id = dbsbuffer_file.workflow
 18  INNER JOIN dbsbuffer_location ON
 19    dbsbuffer_location.id = dbsbuffer_block.location
 20  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 21  AND dbsbuffer_dataset_subscription.subscribed = 1
 22  AND dbsbuffer_block.status = 'Closed'
 23  AND dbsbuffer_block.deleted = 0
 24  GROUP BY dbsbuffer_block.blockname,
 25           dbsbuffer_location.pnn,
 26           dbsbuffer_dataset.path,
 27           dbsbuffer_dataset_subscription.site,
 28           dbsbuffer_workflow.name
 29  HAVING COUNT(*) = SUM(dbsbuffer_workflow.completed);

  COUNT(*) SUM(DBSBUFFER_WORKFLOW.COMPLETED) BLOCKNAME									      PNN	      PATH				       SITE	       NAME
---------- --------------------------------- -------------------------------------------------------------------------------- --------------- ---------------------------------------- --------------- -------------------------------
	 1				   1 /TestEnablesEcalHcal/Tier0_REPLAY_2022-Express-v425/RAW#c0ee5da6-3616-419b-81ba- T0_CH_CERN_Disk /TestEnablesEcalHcal/Tier0_REPLAY_2022-E T2_CH_CERN      Express_Run351572_StreamCalibra
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-SiStripPCLHistos-Express-v425/ALCARECO#e T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /MinimumBias/Tier0_REPLAY_2022-v425/RAW#961b6dcb-7514-4adc-b26c-372792c977fc     T0_CH_CERN_Disk /MinimumBias/Tier0_REPLAY_2022-v425/RAW  T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /L1Accept/Tier0_REPLAY_2022-v425/RAW#575557fc-e92c-45c6-9103-c93572d3538b	      T0_CH_CERN_Disk /L1Accept/Tier0_REPLAY_2022-v425/RAW     T2_CH_CERN      Repack_Run351572_StreamNanoDST_
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-SiPixelCalZeroBias-Express-v425/ALCARECO T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /HcalNZS/Tier0_REPLAY_2022-v425/RAW#32396414-57a4-4188-a6f6-75a208bd4c8f	      T0_CH_CERN_Disk /HcalNZS/Tier0_REPLAY_2022-v425/RAW      T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /NoBPTX/Tier0_REPLAY_2022-v425/RAW#0bbb5f07-c8c2-4725-8251-79c89354e272	      T0_CH_CERN_Disk /NoBPTX/Tier0_REPLAY_2022-v425/RAW       T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /StreamCalibration/Tier0_REPLAY_2022-PromptCalibProdEcalPedestals-Express-v425/A T0_CH_CERN_Disk /StreamCalibration/Tier0_REPLAY_2022-Pro T2_CH_CERN      Express_Run351572_StreamCalibra
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-PromptCalibProdSiStrip-Express-v425/ALCA T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /HLTPhysics/Tier0_REPLAY_2022-v425/RAW#0f75ea20-d421-4946-b366-63655c5bb3a3      T0_CH_CERN_Disk /HLTPhysics/Tier0_REPLAY_2022-v425/RAW   T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /StreamCalibration/Tier0_REPLAY_2022-Express-v425/DQMIO#6082423a-08ad-44b3-8f9e- T0_CH_CERN_Disk /StreamCalibration/Tier0_REPLAY_2022-Exp T2_CH_CERN      Express_Run351572_StreamCalibra

  COUNT(*) SUM(DBSBUFFER_WORKFLOW.COMPLETED) BLOCKNAME									      PNN	      PATH				       SITE	       NAME
---------- --------------------------------- -------------------------------------------------------------------------------- --------------- ---------------------------------------- --------------- -------------------------------
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-SiStripCalZeroBias-Express-v425/ALCARECO T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 4				   4 /StreamExpressCosmics/Tier0_REPLAY_2022-Express-v425/DQMIO#bfdf5c4d-003a-41a1-bd T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /StreamCalibration/Tier0_REPLAY_2022-EcalTestPulsesRaw-Express-v425/ALCARECO#184 T0_CH_CERN_Disk /StreamCalibration/Tier0_REPLAY_2022-Eca T2_CH_CERN      Express_Run351572_StreamCalibra
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-TkAlCosmics0T-Express-v425/ALCARECO#880f T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 5				   5 /ExpressCosmics/Tier0_REPLAY_2022-Express-v425/FEVT#4a678f2c-3945-41cf-b06c-faef T0_CH_CERN_Disk /ExpressCosmics/Tier0_REPLAY_2022-Expres T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /StreamExpressCosmics/Tier0_REPLAY_2022-Express-v425/DQMIO#cec89541-78f7-44a4-aa T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /Cosmics/Tier0_REPLAY_2022-v425/RAW#9d334e74-6687-4670-8f38-f95c720f8287	      T0_CH_CERN_Disk /Cosmics/Tier0_REPLAY_2022-v425/RAW      T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /RPCMonitor/Tier0_REPLAY_2022-v425/RAW#c0abae4e-71a4-4b43-8043-e5e5ee1d6a90      T0_CH_CERN_Disk /RPCMonitor/Tier0_REPLAY_2022-v425/RAW   T2_CH_CERN      Repack_Run351572_StreamRPCMON_T

19 rows selected.

```  

And If I understand your suggestion correctly the change should be something like:
```
SQL> SELECT
  2         count(*),
  3         SUM(dbsbuffer_workflow.completed),
  4         dbsbuffer_block.blockname,
  5         dbsbuffer_location.pnn,
  6         dbsbuffer_dataset.path,
  7         dbsbuffer_dataset_subscription.site,
  8         dbsbuffer_workflow.name
  9  FROM dbsbuffer_dataset_subscription
 10  INNER JOIN dbsbuffer_dataset ON
 11    dbsbuffer_dataset.id = dbsbuffer_dataset_subscription.dataset_id
 12  INNER JOIN dbsbuffer_block ON
 13    dbsbuffer_block.dataset_id = dbsbuffer_dataset_subscription.dataset_id
 14  INNER JOIN dbsbuffer_file ON
 15    dbsbuffer_file.block_id = dbsbuffer_block.id
 16  INNER JOIN dbsbuffer_workflow ON
 17    dbsbuffer_workflow.id = dbsbuffer_file.workflow
 18  INNER JOIN dbsbuffer_location ON
 19    dbsbuffer_location.id = dbsbuffer_block.location
 20  WHERE dbsbuffer_dataset_subscription.delete_blocks = 1
 21    AND dbsbuffer_dataset_subscription.subscribed = 1
 22    AND dbsbuffer_block.status = 'Closed'
 23    AND dbsbuffer_block.deleted = 0
 24    AND dbsbuffer_workflow.completed = 1
 25  GROUP BY dbsbuffer_block.blockname,
 26           dbsbuffer_location.pnn,
 27           dbsbuffer_dataset.path,
 28           dbsbuffer_dataset_subscription.site,
 29           dbsbuffer_workflow.name ;

  COUNT(*) SUM(DBSBUFFER_WORKFLOW.COMPLETED) BLOCKNAME									      PNN	      PATH				       SITE	       NAME
---------- --------------------------------- -------------------------------------------------------------------------------- --------------- ---------------------------------------- --------------- -------------------------------
	 1				   1 /TestEnablesEcalHcal/Tier0_REPLAY_2022-Express-v425/RAW#c0ee5da6-3616-419b-81ba- T0_CH_CERN_Disk /TestEnablesEcalHcal/Tier0_REPLAY_2022-E T2_CH_CERN      Express_Run351572_StreamCalibra
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-SiStripPCLHistos-Express-v425/ALCARECO#e T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /MinimumBias/Tier0_REPLAY_2022-v425/RAW#961b6dcb-7514-4adc-b26c-372792c977fc     T0_CH_CERN_Disk /MinimumBias/Tier0_REPLAY_2022-v425/RAW  T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /L1Accept/Tier0_REPLAY_2022-v425/RAW#575557fc-e92c-45c6-9103-c93572d3538b	      T0_CH_CERN_Disk /L1Accept/Tier0_REPLAY_2022-v425/RAW     T2_CH_CERN      Repack_Run351572_StreamNanoDST_
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-SiPixelCalZeroBias-Express-v425/ALCARECO T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /HcalNZS/Tier0_REPLAY_2022-v425/RAW#32396414-57a4-4188-a6f6-75a208bd4c8f	      T0_CH_CERN_Disk /HcalNZS/Tier0_REPLAY_2022-v425/RAW      T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /NoBPTX/Tier0_REPLAY_2022-v425/RAW#0bbb5f07-c8c2-4725-8251-79c89354e272	      T0_CH_CERN_Disk /NoBPTX/Tier0_REPLAY_2022-v425/RAW       T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /StreamCalibration/Tier0_REPLAY_2022-PromptCalibProdEcalPedestals-Express-v425/A T0_CH_CERN_Disk /StreamCalibration/Tier0_REPLAY_2022-Pro T2_CH_CERN      Express_Run351572_StreamCalibra
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-PromptCalibProdSiStrip-Express-v425/ALCA T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /HLTPhysics/Tier0_REPLAY_2022-v425/RAW#0f75ea20-d421-4946-b366-63655c5bb3a3      T0_CH_CERN_Disk /HLTPhysics/Tier0_REPLAY_2022-v425/RAW   T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /StreamCalibration/Tier0_REPLAY_2022-Express-v425/DQMIO#6082423a-08ad-44b3-8f9e- T0_CH_CERN_Disk /StreamCalibration/Tier0_REPLAY_2022-Exp T2_CH_CERN      Express_Run351572_StreamCalibra

  COUNT(*) SUM(DBSBUFFER_WORKFLOW.COMPLETED) BLOCKNAME									      PNN	      PATH				       SITE	       NAME
---------- --------------------------------- -------------------------------------------------------------------------------- --------------- ---------------------------------------- --------------- -------------------------------
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-SiStripCalZeroBias-Express-v425/ALCARECO T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 4				   4 /StreamExpressCosmics/Tier0_REPLAY_2022-Express-v425/DQMIO#bfdf5c4d-003a-41a1-bd T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /StreamCalibration/Tier0_REPLAY_2022-EcalTestPulsesRaw-Express-v425/ALCARECO#184 T0_CH_CERN_Disk /StreamCalibration/Tier0_REPLAY_2022-Eca T2_CH_CERN      Express_Run351572_StreamCalibra
	 5				   5 /StreamExpressCosmics/Tier0_REPLAY_2022-TkAlCosmics0T-Express-v425/ALCARECO#880f T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 5				   5 /ExpressCosmics/Tier0_REPLAY_2022-Express-v425/FEVT#4a678f2c-3945-41cf-b06c-faef T0_CH_CERN_Disk /ExpressCosmics/Tier0_REPLAY_2022-Expres T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /StreamExpressCosmics/Tier0_REPLAY_2022-Express-v425/DQMIO#cec89541-78f7-44a4-aa T0_CH_CERN_Disk /StreamExpressCosmics/Tier0_REPLAY_2022- T2_CH_CERN      Express_Run351572_StreamExpress
	 1				   1 /Cosmics/Tier0_REPLAY_2022-v425/RAW#9d334e74-6687-4670-8f38-f95c720f8287	      T0_CH_CERN_Disk /Cosmics/Tier0_REPLAY_2022-v425/RAW      T2_CH_CERN      Repack_Run351572_StreamPhysics_
	 1				   1 /RPCMonitor/Tier0_REPLAY_2022-v425/RAW#c0abae4e-71a4-4b43-8043-e5e5ee1d6a90      T0_CH_CERN_Disk /RPCMonitor/Tier0_REPLAY_2022-v425/RAW   T2_CH_CERN      Repack_Run351572_StreamRPCMON_T

19 rows selected.

``` 

The output for a fully completed replay does look equivalent, indeed. Needs to be tested for a running one with a PromptReco paused though.

The difference in those tow queries to me sounds to be the following: 
 * in the former the requirement for matching a `completed=1` requirement for the rows returned is applied  upon aggregation on all of the groups generated by the `group by` statement, 
 * while in the later the condition is applied during the selection statement before the grouping and aggregation. 

This could matter for a running workflow.